services/web/server/src/simcore_service_webserver/login/decorators.py

```diff
@@ -1,28 +1,47 @@
-from functools import wraps
+import functools
+import inspect
 
+from aiohttp import web
 from aiohttp_security.api import check_authorized
-from servicelib.aiohttp.requests_utils import get_request
+from servicelib.aiohttp.typing_extension import HandlerAnyReturn
 from servicelib.request_keys import RQT_USERID_KEY
 
 
-def login_required(handler):
-    """Decorator that restrict access only for authorized users.
+def login_required(handler: HandlerAnyReturn) -> HandlerAnyReturn:
+    """Decorator that restrict access only for authorized users
 
-    User is considered authorized if check_authorized(request) raises no exception
+    - User is considered authorized if check_authorized(request) raises no exception
+    - If authorized, it injects user_id in request[RQT_USERID_KEY]
 
-    Keeps userid in request[RQT_USERID_KEY]
-    """
+    WARNING: Add always @router. decorator FIRST, e.g.
+
+        @router.get("/foo")
+        @login_required
+        async def get_foo(request: web.Request):
+            ...
+
+    and NOT as
+
+        @login_required
+        @router.get("/foo")
+        async def get_foo(request: web.Request):
+            ...
 
-    @wraps(handler)
-    async def wrapped(*args, **kwargs):
-        request = get_request(*args, **kwargs)
+    since the latter will register in `router` get_foo **without** `login_required`
+    """
+    assert set(inspect.signature(handler).parameters.values()) == {  # nosec
+        inspect.Parameter(
+            name="request",
+            kind=inspect.Parameter.POSITIONAL_OR_KEYWORD,
+            annotation=web.Request,
+        )
+    }, f"Expected {handler.__name__} with request as signature, got {handler.__annotations__}"
```

Comment:
If I get it correctly this is a check to enforce what you have written above, right?

Reply:
yes